boa_parser/src/parser/expression/primary/async_function_expression/mod.rs

```diff
@@ -133,7 +142,12 @@ where
             params_start_position,
         )?;
 
-        let function = AsyncFunction::new(name, params, body, has_binding_identifier);
+        let function = AsyncFunction::new(
+            if name.is_some() { name } else { self.name },
+            params,
+            body,
+            name.is_some(),
+        );
```

Comment:
Now that I'm reviewing this I just realized, can't we encode the last parameter into the name itself? An enum like ...

```Rust
enum FunctionName {
    None,
    Direct(Identifier)
    Indirect(Identifier)
}
```

... where `Direct` represents names defined directly on the function definition (`function name()`) and `Indirect` represents names defined by variable assignments and others `let name = function()`.

Reply:
We could but it would only make sense if we would also replace `has_binding_identifier` with that. I think this would be a good improvement for making the ast more readable. I would lift this from this PR, maybe into an issue for new contributers? Seems like a nice change to implement.

Reply:
Yep, gonna open an issue for it.